lib/zipkin-tracer/zipkin_tracer_base.rb

```diff
@@ -23,7 +23,16 @@ def with_new_span(trace_id, name)
 
     def end_span(span)
       span.close
-      if span.annotations.any?{ |ann| ann.value == Annotation::SERVER_SEND }
+      # if in a thread not handling incoming http requests, it will not have Annotation::SERVER_SEND, so the span
+      # will never be flushed and will cause memory leak.
+      # It will have CLIENT_SEND and CLIENT_RECV if the thread sends out http requests, so use CLIENT_RECV as the sign
+      # to flush the span.
+      has_server_recv_span = spans.any? do |s|
+        s.annotations.any? { |ann| ann.value == Annotation::SERVER_RECV }
+      end
+      if span.annotations.any? { |ann| ann.value == Annotation::SERVER_SEND } ||
+          (!has_server_recv_span &&
+              span.annotations.any? { |ann| ann.value == Annotation::CLIENT_RECV })
```

Comment:
how about this logic? it doesn't need more iterating on span annotations, just check if there's previous span. For the leaking case, there's only one span, no previous span.

```
def end_span(span)
      span.close
      # if in a thread not handling incoming http requests, it will not have Annotation::SERVER_SEND, so the span
      # will never be flushed and will cause memory leak.
      # It will have CLIENT_SEND and CLIENT_RECV if the thread sends out http requests, so use CLIENT_RECV as the sign
      # to flush the span.
      has_previous_span = spans.size >= 2
      if (has_previous_span && span.annotations.any? { |ann| ann.value == Annotation::SERVER_SEND }) ||
          (!has_previous_span && span.annotations.any? { |ann| ann.value == Annotation::CLIENT_RECV })
        flush!
        reset
      end
    end
```